lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,76 @@
+'use strict';
+
+const Ajv = require('ajv');
+const schema = require('../configSchema');
+const ajvErrorMessageBuilder = require('../utils/ajvErrorMessageBuilder');
+
+const functionNamePattern = '^[a-zA-Z0-9-_]+$';
+
+class ConfigSchemaHandler {
+  constructor(serverless) {
+    this.serverless = serverless;
+    this.schema = schema;
+
+    deepFreeze(this.schema.properties.service);
+    deepFreeze(this.schema.properties.plugins);
+    deepFreeze(this.schema.properties.resources);
+    deepFreeze(this.schema.properties.package);
+    deepFreeze(this.schema.properties.layers);
+    deepFreeze(this.schema.properties.outputs);
+  }
+
+  validateConfig(userConfig) {
+    const ajv = new Ajv();
+    const validate = ajv.compile(this.schema);
+    validate(userConfig);
+    if (validate.errors) {
+      const messages = ajvErrorMessageBuilder.buildErrorMessages(validate.errors);
+      for (const message of messages) {
+        this.serverless.cli.log(message, 'Serverless', { color: 'red' });
+      }
+    }
+  }
+
+  defineCustomProperty(name, configSchema) {
+    this.schema.properties.custom.properties[name] = configSchema;
+  }
+
+  defineFunctionEvent(name, configSchema) {
+    if (
+      !this.schema.properties.functions.patternProperties[functionNamePattern].properties.events
+        .items.anyOf
```

Comment:
I think it'll be good to have that structure  pre-created in core schema (or is it an issue, to have initially an empty list there ?)

Reply:
If we put it back then the initial schema stored in `lib/configSchema/index.js` would be invalid. AJV performs validation of the schema and does not allow `anyOf` to be an empty array

I don't see the alternative to such implementation. If you have any suggestions, let me know

Reply:
With `runServerless` util, you'd have schema valid (as all AWS plugins will be initialized), you should just test it in context of an AWS service. When not being in service, indeed schema will turn be incomplete, but that's not an issue, as in real world we won't validate anything in such case